src/backend/distributed/connection/connection_management.c

```diff
@@ -333,14 +335,22 @@ CloseNodeConnections(char *nodeName, int nodePort)
 			MultiConnection *connection =
 				dlist_container(MultiConnection, connectionNode, currentNode);
 
-			/* same for transaction state */
-			CloseRemoteTransaction(connection);
-			CloseShardPlacementAssociation(connection);
-
-			/* we leave the per-host entry alive */
-			pfree(connection);
+			connectionList = lappend(connectionList, connection);
 		}
 	}
+
+	/*
+	 * Though it has some overhead, we prefer to accumulate connections into a separate
+	 * list and then close each separately. The reasoning for that is CloseConnection()
+	 * deletes the connection from connectionNode and we don't want to iterate over a
+	 * dlist_node which is modified concurrently in the loop above.
+	 */
+	foreach(connectionCell, connectionList)
```

Comment:
Yes, as far as I can see (and test) we are good. But, I wasn't sure whether is that OK to remove an item from dlist while iterating over it. Is that OK?